File: Java/941. ValidMountainArray.java
class Solution {
    
    public boolean checkASC(int[] A, int e){
        boolean result=false;
        int prev=A[0];
        for(int i=1;i<=e;i++){
            if(prev>=A[i]){
                result=false;
                break;
            }else{
                result=true;
            }
            prev=A[i];
        }
        return result;
    }
    
    public boolean checkDESC(int[] A, int s, int e){
        boolean result=false;
        int prev=A[s];
        if(e-s==1){
            result=true;
        }
        for(int i=s+1;i<e;i++){
            if(prev<=A[i]){
                result=false;
                break;
            }else{
                result=true;
            }
            prev=A[i];
        }
        return result;
    }
    
    
    public boolean validMountainArray(int[] A) {
        int peak=0, prev;
        int len=A.length;
        boolean isMountain=false;
        if(len>=3){
            prev=A[0];
            for(int i=1;i<len;i++){
            if(prev>A[i]){
                peak=i-1;
                break;
            }
            prev=A[i];        
        }
        
        isMountain=(checkASC(A,peak) && checkDESC(A,peak+1,len));
            
        }
             
        return isMountain;
    }
}

File: C++/941.ValidMountainArray.cpp
//Given an array A of integers, return true if and only if it is a valid mountain array.
//Runtime beats 82.32% of cpp submission
//Find the peak(largest number), check whether it is strictly decreasing/increasing from the peak

class Solution {
public:
    bool validMountainArray(vector<int>& A) {
        int j=0,k=0;
        int large=0;
            if(A.size()<3)
                return false;

            for(int i=0;i<A.size()-1;i++){
                if(A[i]<A[i+1])
                    large=i+1;
            }
            if(large==0 ||large==A.size()-1)
                return false;

            while(k<large){
                if(A[k]>=A[k+1])
                    return false;
                k++;
            }
            j=large;
            while(j<A.size()-1){
                if(A[j]<=A[j+1])
                    return false;
                j++;
            }

            return true;

    }
};
